WallsAndHoles/abstractshapebrushtool.cpp

```diff
@@ -0,0 +1,71 @@
+#include "abstractshapebrushtool.h"
+
+AbstractShapeBrushTool::AbstractShapeBrushTool(MapView *mapView, TileMap *tileMap, SharedTileTemplate drawWith)
+    : AbstractTileMapTool(tileMap),
+      mMapView(mapView),
+      mDrawMaterial(drawWith)
+{
+
+}
+
+
+void AbstractShapeBrushTool::cellClicked(int x, int y) {
+    mStartX = x;
+    mStartY = y;
+
+    Q_ASSERT( mStartX >= 0 && mStartX < mTileMap->width() );
+    Q_ASSERT( mStartY >= 0 && mStartY < mTileMap->height() );
+
+    drawOverlay(x, y);
+}
+
+
+void AbstractShapeBrushTool::cellActivated(int x, int y) {
```

Comment:
might be useful if these signals are sent even when outside the tilemap so that you can draw shapes out of bounds...

Don't worry about changing anything for now.

Reply:
The signals are sent from MapView. I think we should add a "mouseReleasedOutOfBounds" signal to MapView.